src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -0,0 +1,103 @@
+            IActionInvokerFactory invokerFactory)
+            : this(actions, invokerFactory, actionContextAccessor: null)
+        {
+        }
+
+        public MvcEndpointDataSource(
+            IActionDescriptorCollectionProvider actions,
+            IActionInvokerFactory invokerFactory,
+            IActionContextAccessor actionContextAccessor)
+        {
+            if (actions == null)
+            {
+                throw new ArgumentNullException(nameof(actions));
+            }
+
+            if (invokerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(invokerFactory));
+            }
+
+            _actions = actions;
+            _invokerFactory = invokerFactory;
+
+            // The IActionContextAccessor is optional. We want to avoid the overhead of using CallContext
+            // if possible.
+            _actionContextAccessor = actionContextAccessor;
+
+            _endpoints = new List<Endpoint>();
+
+            InitializeEndpoints();
+        }
+
+        private void InitializeEndpoints()
+        {
+            // note: this code has haxxx. This will only work in some constrained scenarios
+            foreach (var action in _actions.ActionDescriptors.Items)
+            {
+                if (action.AttributeRouteInfo == null)
+                {
+                    // Action does not have an attribute route
+                    continue;
+                }
+
+                RequestDelegate invokerDelegate = (context) =>
+                {
+                    var values = context.Features.Get<IEndpointFeature>().Values;
+                    var routeData = new RouteData();
+                    foreach (var kvp in values)
+                    {
+                        routeData.Values.Add(kvp.Key, kvp.Value);
+                    }
+
+                    var actionContext = new ActionContext(context, routeData, action);
+                    if (_actionContextAccessor != null)
+                    {
+                        _actionContextAccessor.ActionContext = actionContext;
+                    }
+
+                    var invoker = _invokerFactory.CreateInvoker(actionContext);
+                    return invoker.InvokeAsync();
+                };
+
+                var descriptors = action.FilterDescriptors.OrderBy(f => f, FilterDescriptorOrderComparer.Comparer).Select(f => f.Filter).ToArray();
+                var metadataCollection = new EndpointMetadataCollection(descriptors);
+
+                _endpoints.Add(new MatcherEndpoint(
+                    next => invokerDelegate,
+                    "/" + action.AttributeRouteInfo.Template,
+                    action.RouteValues,
+                    action.AttributeRouteInfo.Order,
+                    metadataCollection,
+                    action.DisplayName));
+            }
+        }
+
+        public override IChangeToken ChangeToken { get; } = NullChangeToken.Singleton;
```

Comment:
This needs to expose a real change token, see:` IActionDescriptorChangeProvider` - you should be able to attach to that.